src/drawbridge.rs

```diff
@@ -193,49 +195,53 @@ pub fn login(
     let token_url = TokenUrl::new(format!("{oidc_domain}oauth/token"))
         .context("Failed to construct token URL")?;
 
-    let client = BasicClient::new(
+    Ok(BasicClient::new(
         ClientId::new(oidc_client_id),
-        None,
+        oidc_client_secret.map(ClientSecret::new),
         auth_url,
         Some(token_url),
     )
     .set_auth_type(AuthType::RequestBody)
-    .set_device_authorization_url(dev_auth_url);
+    .set_device_authorization_url(dev_auth_url))
+}
 
-    let details: StandardDeviceAuthorizationResponse = client
-        .exchange_device_code()
-        .context("Failed to construct device authorization request")?
-        .add_scope(Scope::new("openid".into()))
-        .add_scope(Scope::new("profile".into()))
+pub fn login_client_credentials(
```

Comment:
I think there should only be one `login` function, which should simply be extended with additional parameter (`oidc_client_secret`)

That would reduce duplication and simplify maintenance and would prevent errors, like the one made in this PR already missing e.g. this case https://github.com/enarx/enarx/blob/09bce3fd923bbbb52aac185c5858e377b949f043/src/cli/user/register.rs#L56-L61

Since the list of parameters is getting long, we probably want an `AuthContext`, `OidcContext` or something of that sort be taken as an agument

Reply:
Well, the thing is that there's basically no overlap between this and the current `login` function: the only line of code they share is the `let client = get_oauth2_client(...)`.
So then I'd basically turn both `login` (maybe rename to `login_device_flow`) and `login_client_credentials` into hidden functions, and just have the public `login` function call either of those.
Is that what you'd prefer?

Reply:
> Well, the thing is that there's basically no overlap between this and the current `login` function: the only line of code they share is the `let client = get_oauth2_client(...)`.

That's not true, there's also
```
        .add_scope(Scope::new("manage:drawbridge_users".into()))
        .add_scope(Scope::new("manage:drawbridge_repositories".into()))
        .add_scope(Scope::new("manage:drawbridge_tags".into()))
        .add_extra_param("audience", format!("https://{}/", host))
```
block, for example, which I am much more concerned about, since it's much more likely to get out-of-sync as time goes on and result in a painful-to-debug bugs

Additionally, the `login` and `login_client_credentials` function distinction does not make sense to me.

As a follow-up on https://github.com/enarx/enarx/pull/2354/files#r1026469814
Let's add an enum:
```
enum AuthContext{
   Device{ // fields relevant to device flow },
   Client{ // fields relevant to client cred flow },
}
```
And pass that to `login`

Login can then match on the enum and act accordingly. There's no need to split it into further helper functions

Reply:
So, the problem with merging the scopes and params between the two is that both methods (`oauth2::Client::exchange_client_credentials` and `oauth2::Client::exchange_device_code`) return totally different structs (`oauth2::ClientCredentialsTokenRequest` and `oauth2::DeviceAuthorizationRequest` respectively), which happen to share those methods. So it's not really possible to extract that into common code.
However, I can use the `add_scopes` for both, and just use a statically set list of scopes, which at least forces those two to be in sync.

Reply:
I made that change right now.